python/mxnet/optimizer.py

```diff
@@ -979,8 +979,9 @@ def sync_state_context(self, state, context):
     def set_states(self, states):
         """Sets updater states."""
         states = pickle.loads(states)
-        if isinstance(states, tuple) and len(states) == 2:
-            self.states, self.optimizer = states
+        if isinstance(states, dict):
```

Comment:
?

Reply:
self.states can be length 2 tuple so this is a bug fix.

Reply:
But states are previously a dictionary. This is not backward compatible

Reply:
why would it be a tuple?

Reply:
I missed this myself, it's here: https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/optimizer.py#L375

Reply:
List then?

Reply:
I was confused. `self.states` is actually the dictionary in Updater, so the optimizer's tuple shouldn't be a problem.

Reply:
Ohhhh. Ok, I got confused as well - any particular state may be a tuple, but states is dictionary of them. Sorry for that.